torchdata/datapipes/iter/util/prefetcher.py

```diff
@@ -131,10 +132,129 @@ def reset(self):
 
     def pause(self):
         if self.thread is not None:
+            assert self.prefetch_data is not None
             self.prefetch_data.run_prefetcher = False
 
     def resume(self):
         if self.thread is not None and (
             not self.prefetch_data.stop_iteration or len(self.prefetch_data.prefetch_buffer) > 0
         ):
+            assert self.prefetch_data is not None
             self.prefetch_data.run_prefetcher = True
+
+
+def pin_memory_fn(data, device=None):
+    if hasattr(data, "pin_memory"):
+        return data.pin_memory(device)
+    elif isinstance(data, torch.Tensor):
+        return data.pin_memory(device)
+    elif isinstance(data, str):
+        return data
+    elif isinstance(data, collections.abc.Mapping):
+        pinned_data = {k: pin_memory_fn(sample, device) for k, sample in data.items()}
+        try:
+            return type(data)(**pinned_data)
+        except TypeError:
+            # The mapping type may not support `__init__(iterable)`.
+            return pinned_data
+    elif isinstance(data, collections.abc.Sequence):
+        pinned_data = [pin_memory_fn(sample, device) for sample in data]  # type: ignore[assignment]
+        try:
+            type(data)(*pinned_data)
+        except TypeError:
+            # The sequence type may not support `__init__(iterable)` (e.g., `range`).
+            return pinned_data
+    else:
```

Comment:
Special handling for tuples and namedtuples too?

Reply:
I don't think we need that. The original implementation in PyTorch Core is for BC. TBH, we don't need to have an aligned implementation in TorchData.
If users want to use the original `pin_memory` functino, they can reference it and give `PinMemory` DataPipe